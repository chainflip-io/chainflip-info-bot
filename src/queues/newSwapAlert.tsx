import { renderToStaticMarkup } from 'react-dom/server';
import { type JobConfig, type DispatchJobArgs, type JobProcessor } from './initialize.js';
import { Bold, ExplorerLink, Line, TokenAmount, UsdValue } from '../channels/formatting.js';
import { platforms } from '../config.js';
import { humanFriendlyAsset } from '../consts.js';
import getSwapInfo from '../queries/getSwapInfo.js';
import baseLogger from '../utils/logger.js';

const name = 'newSwapAlert';
type Name = typeof name;

const logger = baseLogger.child({ queue: name });

type Data = {
  swapRequestId: `${number}`;
};

declare global {
  interface JobData {
    [name]: Data;
  }
}

type SwapInfo = Awaited<ReturnType<typeof getSwapInfo>>;

const buildMessageData = ({
  swapInfo,
}: {
  swapInfo: SwapInfo;
}): Extract<DispatchJobArgs, { name: 'messageRouter' }>[] =>
  platforms.map((platform) => {
    const message = renderToStaticMarkup(
      <>
        <Line>
          New swap detected:{' '}
          <Bold platform={platform}>
            <ExplorerLink platform={platform} path={`/swaps/${swapInfo.requestId}`} prefer="link">
              #{swapInfo.requestId}
            </ExplorerLink>
          </Bold>
        </Line>
        <Line>
          Swapping <Bold platform={platform}>{humanFriendlyAsset[swapInfo.sourceAsset]}</Bold> for{' '}
          <Bold platform={platform}>{humanFriendlyAsset[swapInfo.destinationAsset]}</Bold>
        </Line>
        <Line>
          ðŸ“¥{' '}
          <Bold platform={platform}>
            <TokenAmount amount={swapInfo.depositAmount} asset={swapInfo.sourceAsset} />
          </Bold>
          <UsdValue amount={swapInfo.depositValueUsd} />
        </Line>
        {swapInfo.depositAddress && (
          <Line>
            Deposit address: <Bold platform={platform}>{swapInfo.depositAddress}</Bold>
          </Line>
        )}
        <Line>
          Destination address: <Bold platform={platform}>{swapInfo.destinationAddress}</Bold>
        </Line>
        {swapInfo.transactionRefs.length > 0 && (
          <>
            <Line>Transaction refs:</Line>
            {swapInfo.transactionRefs.map(({ ref, chain }) => (
              <Line key={ref}>
                {ref}{' '}
                <ExplorerLink key={ref} path={ref} chain={chain} platform={platform} prefer="text">
                  [view]
                </ExplorerLink>
              </Line>
            ))}
          </>
        )}
      </>,
    ).trimEnd();

    return {
      name: 'messageRouter' as const,
      data: {
        platform,
        message,
        filterData: { name: 'NEW_SWAP', usdValue: swapInfo.depositValueUsd?.toNumber() || 0 },
        opts: { disablePreview: true },
      },
    };
  });

const processJob: JobProcessor<Name> = (dispatchJobs) => async (job) => {
  logger.info(job.data, 'Alerting about new swap request');

  const swapInfo = await getSwapInfo(job.data.swapRequestId);

  if (swapInfo.freshness !== 'stale') {
    await dispatchJobs(buildMessageData({ swapInfo }));
  } else {
    logger.warn({ swapRequestId: job.data.swapRequestId }, 'skipping stale job');
  }
};

export const config: JobConfig<typeof name> = {
  name,
  processJob,
};
