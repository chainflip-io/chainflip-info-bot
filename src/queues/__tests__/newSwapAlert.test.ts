import { describe, it, expect, vi, beforeEach, afterEach } from 'vitest';
import { config } from '../newSwapAlert.js';

describe('newSwapAlert', () => {
  beforeEach(() => {
    vi.useFakeTimers();
  });

  afterEach(() => {
    vi.useRealTimers();
  });

  describe('processJob', () => {
    it('dispatches a job to send a message', async () => {
      vi.setSystemTime(new Date('2024-10-25T12:42:30+00:00'));

      const dispatchJobs = vi.fn();

      await config.processJob(dispatchJobs)({
        data: {
          swapRequestId: '77697',
        } as JobData['newSwapAlert'],
      } as any);

      expect(dispatchJobs.mock.lastCall).toMatchSnapshot();
    });
  });

  it('skips stale swaps', async () => {
    const dispatchJobs = vi.fn();
    await config.processJob(dispatchJobs)({
      data: {
        swapRequestId: '77697',
      } as JobData['newSwapAlert'],
    } as any);

    expect(dispatchJobs).not.toHaveBeenCalled();
  });
});
