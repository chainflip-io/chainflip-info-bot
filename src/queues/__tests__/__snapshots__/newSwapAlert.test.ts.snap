// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`newSwapAlert > processJob > dispatches a job to send a message 1`] = `
[
  [
    {
      "data": {
        "filterData": {
          "name": "NEW_SWAP",
          "usdValue": 5611.3799573378,
        },
        "message": "New swap detected: <strong><a href="https://scan.chainflip.io/swaps/77697">#77697</a></strong>
Swapping <strong>FLIP on Ethereum</strong> for <strong>USDT on Ethereum</strong>
ðŸ“¥ <strong>5,000 FLIP on Ethereum</strong> ($5,568.07)
Deposit address: <strong>0xf05914614bc56915ce5dac12d8eabdbd4bf1e470</strong>
Destination address: <strong>0x4c632678b9a2c2f0ca1472d17c683aab6dc16523</strong>
Transaction refs:
0x17077234a0e19f7b4ab2d2491c09065d8a164199ffbbd9fd02078f8394eb36e4 <a href="https://etherscan.io/tx/0x17077234a0e19f7b4ab2d2491c09065d8a164199ffbbd9fd02078f8394eb36e4">[view]</a>",
        "platform": "telegram",
      },
      "name": "messageRouter",
    },
    {
      "data": {
        "filterData": {
          "name": "NEW_SWAP",
          "usdValue": 5611.3799573378,
        },
        "message": "New swap detected: **[#77697](https://scan.chainflip.io/swaps/77697)**
Swapping **FLIP on Ethereum** for **USDT on Ethereum**
ðŸ“¥ **5,000 FLIP on Ethereum** ($5,568.07)
Deposit address: **0xf05914614bc56915ce5dac12d8eabdbd4bf1e470**
Destination address: **0x4c632678b9a2c2f0ca1472d17c683aab6dc16523**
Transaction refs:
0x17077234a0e19f7b4ab2d2491c09065d8a164199ffbbd9fd02078f8394eb36e4 [[view]](https://etherscan.io/tx/0x17077234a0e19f7b4ab2d2491c09065d8a164199ffbbd9fd02078f8394eb36e4)",
        "platform": "discord",
      },
      "name": "messageRouter",
    },
    {
      "data": {
        "filterData": {
          "name": "NEW_SWAP",
          "usdValue": 5611.3799573378,
        },
        "message": "New swap detected: https://scan.chainflip.io/swaps/77697
Swapping FLIP on Ethereum for USDT on Ethereum
ðŸ“¥ 5,000 FLIP on Ethereum ($5,568.07)
Deposit address: 0xf05914614bc56915ce5dac12d8eabdbd4bf1e470
Destination address: 0x4c632678b9a2c2f0ca1472d17c683aab6dc16523
Transaction refs:
0x17077234a0e19f7b4ab2d2491c09065d8a164199ffbbd9fd02078f8394eb36e4 [view]",
        "platform": "twitter",
      },
      "name": "messageRouter",
    },
  ],
]
`;
